midend/interpreter.cpp

```diff
@@ -310,33 +310,50 @@ void SymbolicStruct::dbprint(std::ostream& out) const {
 }
 
 SymbolicHeaderUnion::SymbolicHeaderUnion(const IR::Type_HeaderUnion* type,
-                               bool uninitialized,
-                               const SymbolicValueFactory* factory) :
-        SymbolicStruct(type, uninitialized, factory),
-        valid(new SymbolicBool(false)) {}
+                                         bool uninitialized,
+                                         const SymbolicValueFactory* factory) :
+        SymbolicStruct(type, uninitialized, factory) {}
+
+void SymbolicHeaderUnion::setValid(bool v, cstring field) {
+    if (field) {
+        fieldValue[field]->to<SymbolicHeader>()->setValid(v);
+        for (auto f : type->to<IR::Type_StructLike>()->fields) {
+            if (f->name.name != field)
+                fieldValue[f->name.name]->setAllUnknown();
+        }
+    }
+}
 
-void SymbolicHeaderUnion::setValid(bool v) {
-    if (!v)
-        setAllUnknown();
-    valid = new SymbolicBool(v);
+SymbolicBool* SymbolicHeaderUnion::isValid() const {
+    for (auto f : type->to<IR::Type_StructLike>()->fields) {
+         auto fieldsClone = fieldValue;
+         if (fieldsClone[f->name.name]) {
+             if (auto fildValid = fieldsClone[f->name.name]->to<SymbolicHeader>()->valid) {
+                 if (fildValid->isKnown())
+                     return new SymbolicBool(fildValid->value);
+             }
+         }
+    }
+    return new SymbolicBool(false);
 }
 
 SymbolicValue* SymbolicHeaderUnion::get(const IR::Node* node, cstring field) const {
-    if (valid->isKnown() && !valid->value)
-        return new SymbolicStaticError(node, "Reading field from invalid header union");
+    if (this->isValid()->isKnown()  && !this->isValid()->value)
```

Comment:
is this line really necessary?

Reply:
I'm not even sure, can we read header from unvalid headerUnion?

Reply:
The language spec definitely says that for a normal header, if you attempt to use one of its fields while it is invalid, then the "read" of the field is allowed to return any arbitrary value (basically, giving free reign to the implementation for whatever it wants to do).  The language spec should say somewhere that this is also the case for header unions, but I have not scanned the language spec for that question specifically on header unions today.  If you look for it and do not find any such mention, let me know.  It should be in this section: https://p4.org/p4-spec/docs/P4-16-v1.2.2.html#sec-uninitialized-values-and-writing-invalid-headers

Reply:
Reading from that section...

> Where a header is mentioned, it may be a member of a header_union, an element in a header stack, or a normal header.

Reply:
I relied on the Header implementation here https://github.com/vladyslav-dubina/p4c/blob/b1034c8742c5d301913584b0be59291cc5575251/midend/interpreter.cpp#L342
And tried to make similar behavior. I don't quite understand what's wrong.

Reply:
I added the extract_for_header_union.p4 test, it shows the ability to read a field from an invalid header union , and it prints a warning about reading from an invalid header union